[Mobile] Add external link inline component

enejb opened the pull request:
## Description
Add inline link to an external resource in the mobile app. It replaces the current `FooterMessageLink` Component since it does the same thing as the external link to create consistency with the Gutenberg components code base.

## How has this been tested?
1. Open the PR on a mobile device. Start a mobile dev environment. 
2. Open the editor and add the heading block. 
3. Open the block settings notice the new inline link. 

## Screenshots <!-- if applicable -->
See Header Settings

<img width="404" alt="Screen Shot 2021-07-05 at 2 30 51 PM" src="https://user-images.githubusercontent.com/115071/124519690-e024a000-dd9e-11eb-9d28-92aa779cfca2.png">

Image block Alt Text settings.
<img width="395" alt="Screen Shot 2021-07-05 at 2 41 39 PM" src="https://user-images.githubusercontent.com/115071/124519770-1eba5a80-dd9f-11eb-8092-e208c7b0f82a.png">


## Types of changes
Enhancement. 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

enejb committed:
Simply External Link component

enejb committed:
Update the Anchor Hook to use ExternalLink

Add a space character to give the link some space.

enejb committed:
Remove FooterMessageLink and replace it with ExternalLink

dcalhoun commented on packages/block-editor/src/hooks/anchor.js:
```diff
@@ -77,17 +77,16 @@ export const withInspectorControl = createHigherOrderComponent(
 										{ __( 'Learn more about anchors' ) }
 									</ExternalLink>
-								) }
+								}
```
The wrapping `{ }` is unnecessary here, since we rendering a JSX element inside. I.e. we can render `<ExternalLink />` instead of `{<ExternalLink />}`.

dcalhoun commented on packages/components/src/external-link/index.native.js:
```diff
@@ -2,23 +2,32 @@
+			style={ externalLink }
 			onPress={ () => Linking.openURL( href ) }
 			accessibilityLabel={ __( 'Open link in a browser' ) }
 		>
```
While switching from a `Touchable` to a `Text` fixes the visual issue, I wanted to call out that the accessibility of this component is very poor currently. 

First, there is no visual feedback when tapping the inline link. Ideally we provide immediate feedback when a user taps an element. By moving from a `Touchable` to a `Text` we lose the built-in on press feedback. We could potentially remedy this by toggling a custom style via state while the `Text` is actively pressed. 

Second, the `accessibilityLabel` for this component is not reachable and never read aloud. From a quick search, we might be able to improve this situation by [using a conditional `onPress` for screen readers](https://callstack.com/blog/react-native-android-accessibility-tips/) for inline text links.

dcalhoun commented on packages/components/src/mobile/bottom-sheet-text-control/README.md:
```diff
@@ -34,10 +34,9 @@ const MyTextControl = () => {
-							href={ 'https://wordpress.org/' }
-							value={ __( 'Visit WordPress.org' ) }
-						/>
+						<ExternalLink href={ 'https://wordpress.org/' } >
+							{ __( 'Visit WordPress.org' ) }
+						</ExternalLink>
```
While I support removing `FooterMessageLink`, as it seems duplicative of `ExternalLink` and the work in this PR addresses the visual issue, I want to call out that by doing so we are further spreading the aforementioned accessibility issues it possess into this `BottomSheetTextControl` component. I.e. now any use of `BottomSheetTextControl` will suffer from these issues.

From a quick debugging session, I identified one preexisting issue specifically (in addition to the aforementioned issue mentioned in a earlier comment). `footerNote` allows a React element, which works for visual rendering, but its value is ultimately passed as the `help` prop to `BottomSheet.Cell`, which attempts to render it as a string for the accessibility label. This results in "Object Object" being read aloud by the screen reader.  

I recognize some of the accessibility issues existed before this PR, but I wanted to call it out for visibility.

SiobhyB replied:
> This results in "Object Object" being read aloud by the screen reader.

I believe this was fixed via commits that Enej made following this feedback, which involved refactoring `footerNote` and replacing it with `help`. In my testing `Object Object` is read aloud when `trunk` is checked out, but the full message is read when this branch is checked out:

| Trunk  | This Branch |
| ------------- | ------------- |
| <video src="https://user-images.githubusercontent.com/2998162/190619298-dfac4b5d-37ab-4de9-b4cf-d5b0419769a5.mov" /> | <video src="https://user-images.githubusercontent.com/2998162/190621249-139f3ee5-c80a-43ef-bc21-c8c9aee94a2d.mov" /> |

dcalhoun reviewed (approved):
Thank you for putting this together! From a visual standpoint, this looks good to me. I tested on an iPhone 12 mini Simulator and Samsung Galaxy S20. Removing `FooterMessageLink` is also a really good idea IMO. 

I left several comments regarding accessibility concerns. I do understand some of them preexisted, so if you would like to ship this visual fix as-is with the accessibility issues, I understand and support that. However, we should at least create a new issue for the accessibility issues if they do not already exist. WDYT? 

If you would prefer me to create the new issues, just let me know and I can. Full transparency — it'll probably be a few days until I find the time to create them.

SiobhyB commented on packages/components/src/mobile/bottom-sheet/footer-message-link/README.md:
```diff
@@ -1,37 +0,0 @@
-# Footer Message Link
-
-FooterMessageLink allows for adding a link within a FooterMessageControl component.
```
I noticed [the README for the ExternalLink component](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/external-link/README.md) is pretty barebones. That isn't a big problem by any measure, given the component itself is simple and self-explanatory, but it may be a nice "extra touch" to add a description at the top of the README. Something like this might be helpful for giving quicker clarity to people scanning through the components.  

```
Renders a hyperlink that, when tapped or clicked, opens in a new window. 
```

There are also a couple of differences between the web and native props. Perhaps an extra `props` section to the README could be helpful? A suggestion:

```
## Props

The text that's displayed within the component's tags will be "clickable" and link to any URL provided via the `href` prop. As outlined below, the `href` prop is the only required prop. 

There are two other optional props that are web only. That is, they're not available for use with the React Native implementation of the editor.

### `href`

The URL that is being linked to and will open in a new window when tapped or clicked.

-   Type: `String`
-   Required: Yes

### `className` (web only)

An optional CSS class that's added to the hyperlink's HTML. 

-   Type: `String`
-   Required: No

### `rel` (web only)

A <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel">rel attribute</>.

-   Type: `String`
-   Required: No
```
Let me know your thoughts!

enejb replied:
Thanks for the great suggestion @SiobhyB.

SiobhyB reviewed (commented):
Thank you for looping me in here @enejb! I hadn't spotted the `<ExternalLink>` component and agree it makes a lot more sense to re-use it vs. the `<FooterMessageLink>` component.

I tested on a Pixel 3 emulator and confirm that everything looks good visually to me. I was able to replicate the `Object Object` screenreader problem that @dcalhoun reported, however. Let me know if you'd like me to create the GitHub issue for that, as I appreciate it was ultimately me that introduced that bug with the way I set up the `<FooterMessageLink>` component.

The only additional suggestion I had was around expanding on the current README for `ExternalLink`. That's a very small suggestion, but may be useful for clarity and might help to make the component more discoverable for folks quickly browsing through.

enejb committed:
Update external link readme

enejb committed:
Add New HelpText Component.

enejb committed:
Include the Help Text in components

enejb committed:
Update BottomSheet text control

enejb committed:
Add accessibilityRole and accessibilityHint to ExternalLink

enejb committed:
Update Image Edit to use HelpText

enejb committed:
Move Help Text

enejb committed:
Update the anchor to use HelpText

SiobhyB reviewed (approved):
Circling back to make sure I "approve" this following the changes. LGTM. :) 🙇

enejb committed:
Fix the url link

enejb committed:
Minor fix

enejb committed:
Update the ImageEdit to use HelpText

SiobhyB commented:
Noting for posterity that this PR is going to be worked on within the next few month or so.

SiobhyB committed:
Merge branch into trunk

SiobhyB committed:
Adapt code to work with new Typescript component

As the web's version of the ExternalLink component has been converted to Typescript, it's necessary to make some tweaks to the new HelpText component (which imports ExternalLink). This commit does the following:

- Fixes the path ExternalLink is imported from.
- Declares specific types for each prop.

SiobhyB committed:
Fix lint error: Remove whitespace in string

SiobhyB committed:
Revert accidental code deletion

SiobhyB committed:
Update reference to old/removed component

As 'FooterMessageLink' has been replaced with 'ExternalLink', this commit updates an outdated reference to the now-removed component.

SiobhyB committed:
Add necessary space between text and URL

As we'll always need a space between text and the URL that's returned by the HelpText component, this commit goes ahead to add the space within the component itself. This will make the component easier to use "out of the box", as it won't be necessary for developers to remember.

SiobhyB committed:
Add necessary space between text and link

As with the previous commit, the necessary space between text and link in the HelpText component is added here.

SiobhyB committed:
Update README for ExternalLink for accuracy

This README has been updated since this PR was first worked on (in https://github.com/WordPress/gutenberg/pull/41681/). As such, this commit brings in the latest updates for accuracy.

SiobhyB committed:
Only add space if there is *both* text and a URL

SiobhyB committed:
Apply correct style to HelpText component

SiobhyB committed:
Refine native styles to match FooterControl

SiobhyB committed:
Tweak HelpText styling to match FooterControl

SiobhyB committed:
Refactor FooterMessageControl in LinkSettings

SiobhyB committed:
Introduce a top border for the HelpText component

SiobhyB commented:
I've gone ahead to update this branch and made some tweaks to resolve errors against `trunk`. I believe all concerns related to screen reader compatibility have now been resolved, with VoiceOver correctly reading all text using the updated component in my testing. If there are any further accessibility concerns I'm missing, I'd be happy to create a separate GitHub issue.

A noteworthy change is that I updated the `LinkSettings` component [to use the new `HelpText` component](https://github.com/WordPress/gutenberg/blob/3f092cfbc4b6e184f6d0ffd8764c1ad2e8ace0ec/packages/components/src/mobile/link-settings/index.native.js#L308-L314), rather than `FooterMessageControl`. This was to account for the fact that `FooterMessageLink` was being used [in `embed-link-settings.native.js`](https://github.com/WordPress/gutenberg/pull/33210/files#diff-e5ac350e88730a8a1b363d454556cb3f6b966945da15d38be9b9280f484ba368). I've taken before vs. after screenshots of the two screens this change impacts, to verify that it doesn't result in a UI change.

<details>
 <summary>Social Icons Link Setting ⤵</summary>

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/190682694-aacaa461-dd04-45f6-b2ab-a59cd9ff4ed4.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/190683059-8d76c3cd-357e-48c4-96a2-91de33e688e7.png" width="100%"> |
</details>

<details>
 <summary>Empty Embed Block's Link Setting ⤵</summary>

| Before | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/190686276-48c67cb2-7d7f-434b-ba02-4736725a0bbe.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/190701788-e40e20f5-960d-4308-a453-ec8f82c675b1.png" width="100%"> |
</details>

@dcalhoun, I've gone ahead to request your reviews due to your previous involvement on this PR, but let me know if you're not available or if you'd rather someone else take the review. Thanks! 🙇‍♀️

dcalhoun commented:
> @dcalhoun, I've gone ahead to request your reviews due to your previous involvement on this PR, but let me know if you're not available or if you'd rather someone else take the review. Thanks! 🙇‍♀️

Hey @SiobhyB. 👋🏻 I have begun reviewing this again, but I will likely need a week or more to complete my review as I will be AFK for a bit. I plan to follow up middle of next week, but feel free to seek other help if you'd prefer.

SiobhyB commented:
> I have begun reviewing this again, but I will likely need a week or more to complete my review as I will be AFK for a bit. I plan to follow up middle of next week, but feel free to seek other help if you'd prefer.

Thank you, David! That sounds good, I don't think there's an urgency to this PR. 🙇‍♀️

dcalhoun commented on packages/components/src/external-link/index.native.js:
```diff
@@ -2,23 +2,34 @@
+			style={ externalLink }
 			onPress={ () => Linking.openURL( href ) }
-			accessibilityLabel={ __( 'Open link in a browser' ) }
+			accessibilityLabel={ children }
```
To avoid future a11y label issues, it may be worth throwing an error in this component if `typeof children !== 'string'`. WDYT?

SiobhyB replied:
Agreed that's a good idea, I've added an error message in https://github.com/WordPress/gutenberg/pull/33210/commits/8d4d17e44196b318e866d9042fddd6ca9c4b32f8.

dcalhoun commented on packages/block-editor/src/hooks/anchor.js:
```diff
@@ -97,6 +85,23 @@ export const withInspectorControl = createHigherOrderComponent(
 						autoComplete="off"
+						separatorType="none"
+						help={
+							isWeb && (
+								<>
+									{ __(
+										'Enter a word or two — without spaces — to make a unique web address just for this block, called an “anchor.” Then, you’ll be able to link directly to this section of your page.'
+									) }
+									<ExternalLink
+										href={
+											'https://wordpress.org/support/article/page-jumps/'
+										}
+									>
+										{ __( 'Learn more about anchors' ) }
+									</ExternalLink>
+								</>
+							)
```
🤔 This comment is quite broad and provides more questions than solutions. I understand if we'd prefer to move forward without addressing this. 

---

Here the web now relies upon passing `help` to a robust, composable `TextControl` → `BaseControl` component that manages help text and links; native now relies upon `HelpText` composed further along in this hook file rather than the previous approach of passing `help` down to `TextControl` → `Cell`. 

While native's `Cell` is quite complex and overly far-reaching — it is quite different from the web's composable approach in `BaseControl` and likely not the best abstraction long term — I am hesitant move the help text from `TextControl` → `Cell` to inline `HelpText` in this hook as it further diverges web and native component design, which is unfortunate. 

My interpretation for the rationale of this change is that we cannot pass a multiple children types — string and interactive link — on native. Doing so leads to the `Object Object` [accessibility (a11y) label issue](https://github.com/WordPress/gutenberg/pull/33210#discussion_r664627933). My hope is that we could identify a way to address the a11y label issue while also avoiding the divergence from web's `help` property here. 

An example of a possible approach might involve flattening `help` props rather than introducing the `HelpText` component. The flattened `help` prop could be used for a11y labels, the original `help` prop could be rendered as normal for visible UI. Note: I didn't test any performance of such code.  

```js
const helpPropExample = (
  <>
    A top-level sentence.{" "}
    <ExternalLink href={"https://wordpress.org/support/article/page-jumps/"}>
      A nested link
    </ExternalLink>
  </>
);

function flattenElementToString(reactElement = "") {
  if (!reactElement) {
    return;
  }
  const children = reactElement.props?.children || reactElement;

  if (Array.isArray(children)) {
    return children.reduce(
      (acc, child) => acc + flattenElementsToString(child),
      ""
    );
  } else if (React.isValidElement(children)) {
    return flattenElementsToString(children);
  }

  return children;
}

const flatHelpPropExample = flattenElementToString(helpPropExample);
console.log(">>>", flatHelpPropExample); // >>> A top-level sentence. A nested link

<Text accessibilityLabel={flatHelpPropExample}>{helpPropExample}</Text>;
```

dcalhoun commented on packages/block-library/src/embed/embed-link-settings.native.js:
```diff
@@ -35,15 +32,9 @@ const EmbedLinkSettings = ( {
-			),
+		help: {
+			url: __( 'https://wordpress.org/support/article/embeds/' ),
+			moreLinkText: __( 'Learn more about embeds' ),
```
🤔 This comment is quite broad and provides more questions than solutions. I understand if we'd prefer to move forward without addressing this. 

---

The replacement of the composable approach with a prop-driven approach is less than ideal IMO. I.e. the ability to pass React elements means we have more optionality in regards to what we pass without defining additional props. The prop-driven approach limits us to a very specific composition defined within `HelpText`. 

For this specific example of `help`, this ultimate uses `HelpText`, but only passes props for a "more link." So, in reality, this only needs to render an `ExternalLink`, not an entire `HelpText`. This unnecessary complexity is less than ideal IMO. 

My understanding for the rationale for this change is similar to my previous comment, it is difficult to pass both a string and a interactive link and not run into accessibility label issues.

dcalhoun commented on packages/components/src/help-text/README.md:
```diff
@@ -0,0 +1,31 @@
+
+## Props
+
+The text that's displayed within the component's tags will be "clickable" and extenal link will be added at the end.
```
This statement feels out of place underneath the heading of "Props." Was this intended to be a description of the `children` prop? 

Also, I believe this only to be true on native, as it currently applies an `onPress` to a wrapping `Text` element. The web version merely renders the child text. I wonder if we can better describe this.

dcalhoun commented on packages/components/src/help-text/README.md:
```diff
@@ -0,0 +1,31 @@
+# HelpText
+
+Renders a help text shell with a ExternalLink appended at the end.
```
```suggestion
Renders help text with a `ExternalLink` appended at the end of the text.
```

dcalhoun commented on packages/components/src/help-text/README.md:
```diff
@@ -0,0 +1,31 @@
+
+### `url`
+
+An optional URL that is being linked to and will open in a new window when tapped or clicked.
```
```suggestion
An optional URL to link, opens in a new window when tapped or clicked.
```

dcalhoun commented on packages/components/src/help-text/README.md:
```diff
@@ -0,0 +1,31 @@
+
+### `moreLinkText` 
+
+An optional Text as the click abel text at the end.
```
```suggestion
An optional text string to display as the link label for the optional `url` prop. 
```

dcalhoun commented on packages/components/src/help-text/index.js:
```diff
@@ -0,0 +1,24 @@
+ * @param {import('react').ReactNode} props.children
+ * @param {string}                    props.url
+ */
+const HelpText = ( { moreLinkText, children, url } ) => {
```
🤔 This comment is quite broad and provides more questions than solutions. I understand if we'd prefer to move forward without addressing this. 

---

This is the first time I have encountered us adding a web component. That may be OK, but it is a bit odd to add it when there is no usage of it anywhere. 

For the most part, I've seen the web project lead, and the native project sometimes splits existing web components. Additionally, there is a `packages/components/src/mobile/` directory housing mobile-only components for specific mobile UI patterns that, I believe, arrived during the [mono-repo merge](https://github.com/WordPress/gutenberg/pull/22714).

I raise this primarily as it is largely uncharted territory for me. WDYT?

dcalhoun commented on packages/components/src/help-text/index.native.js:
```diff
@@ -0,0 +1,61 @@
+					accessibilityHint={
+						url ? __( 'Opens link in a browser' ) : null
+					}
+					onPress={ () => {
+						AccessibilityInfo.isScreenReaderEnabled().then(
+							( enabled ) => enabled && Linking.openURL( url )
+						);
+					} }
```
Similar to `accessibilityRole` and `accessibilityHint`, I wonder if this should be conditionally set if there is a `url` defined. 

```js
onPress={ url ? () => { /* [...] */ } : undefined }
```

dcalhoun commented on packages/components/src/help-text/index.native.js:
```diff
@@ -0,0 +1,61 @@
+				>
+					{ children }
+					{ children && url && ' ' }
+					{ url && (
```
To avoid the possibility of a link without text, I wonder if we should create a compound conditional for all checks of `url` in this file. E.g.:

```js
const hasMoreLink = url && moreLinkText
// [...]
{ children && hasMoreLink && ' ' }
{ hasMoreLink && ( /* [...] */ ) }
```

SiobhyB replied:
I appreciate that thought, gone ahead to implement your suggestion in https://github.com/WordPress/gutenberg/pull/33210/commits/8ee995c3931ed809af24964bff4d9d7764881ac0.

dcalhoun commented on packages/components/src/index.native.js:
```diff
@@ -98,7 +99,6 @@ export { default as ColorSettings } from './mobile/color-settings';
 export { default as SegmentedControls } from './mobile/segmented-control';
 export { default as FocalPointSettingsPanel } from './mobile/focal-point-settings-panel';
 export { default as BottomSheetTextControl } from './mobile/bottom-sheet-text-control';
-export { default as FooterMessageLink } from './mobile/bottom-sheet/footer-message-link/footer-message-link';
```
`/mobile/bottom-sheet/footer-message-link/footer-message-link` appears to still be in place. I presume we should remove this file as well, correct?

SiobhyB replied:
Good catch, thank you! I've gone ahead to delete that file in https://github.com/WordPress/gutenberg/pull/33210/commits/79ea8b57d32975d80a06ec38eb569ea8b8baa2db.

dcalhoun commented on packages/components/src/help-text/index.native.js:
```diff
@@ -0,0 +1,61 @@
+			) }
+			<View style={ styles[ 'help-text' ] }>
+				<Text
+					accessibilityRole={ url ? 'link' : 'text' }
+					accessibilityHint={
+						url ? __( 'Opens link in a browser' ) : null
+					}
```
This combination of props along with the children text appear to work for iOS. However, it does not appear to read any label loud for Android. 

From testing briefly, a quick fix is adding `accessibilityLabel={ moreLinkText }`. It causes the `moreLinkText` to be read for both platforms. Unfortunately, it no longer reads the entire help text aloud on iOS. 

Using something like `accessibilityLabel={ children + moreLinkText }` is not possible as it reads aloud "Object Object," presumably because of translation string methods used: `__()`.

dcalhoun commented on packages/components/src/mobile/bottom-sheet-text-control/README.md:
```diff
@@ -31,13 +31,12 @@ const MyTextControl = () => {
 					<>
 						{ __( 'A footer note to add to the component! ' ) }
-						<FooterMessageLink
-							href={ 'https://wordpress.org/' }
-							value={ __( 'Visit WordPress.org' ) }
-						/>
+						<ExternalLink href={ 'https://wordpress.org/' } >
+							{ __( 'Visit WordPress.org' ) }
+						</ExternalLink>
 					</>
 				}
```
While the original "Object Object" a11y label issue appears to be resolved in the app, I believe example offered here would still result in a problematic a11y label. We likely need to replace the example showcasing an inline link with the usage of `HelpText` [showcased in the changes this PR makes to the Image edit component](https://github.com/WordPress/gutenberg/blob/52d62bcc6a4f21363f48e86e5d627ff1ac9e8532/packages/block-library/src/image/edit.native.js#L576-L587) or change the entire architecture of this PRs approach to composing help texts with links.

dcalhoun reviewed (commented):
Hey @SiobhyB. 👋🏻 I appreciate your patience. Returning to this PR was challenging to regain the context required to provide (hopefully) valuable feedback. 

> Add inline link to an external resource in the mobile app. It replaces the current `FooterMessageLink` Component since it does the same thing as the external link to create consistency with the Gutenberg components code base.

My review is quite long. I hope it is not overwhelming. I felt this type of refactor introducing new paradigms to the code base was worthy of a long look. The scope of this PR seems to have grown past the original goal of replacing `FooterMessageLink` with an external link component, presumably to address the accessibility label issues. 

Some comments call out specific issues we should address before merging. However, other of the comments are pretty open-ended. As mentioned in those opened-ended comments, it is fine if we'd prefer to move forward with the current approach instead. 

Thanks!

SiobhyB committed:
Update README description for clarity

Co-authored-by: David Calhoun <438664+dcalhoun@users.noreply.github.com>

SiobhyB committed:
Update README description for clarity

Co-authored-by: David Calhoun <438664+dcalhoun@users.noreply.github.com>

SiobhyB committed:
Update README description for clarity

Co-authored-by: David Calhoun <438664+dcalhoun@users.noreply.github.com>

SiobhyB committed:
Delete unused 'footer-message-link' file

SiobhyB committed:
Help component: Avoid rendering link without text

A conditional has been created to check that s link has text associated with it before rendering.

SiobhyB committed:
Throw error if wrong data type is passed to link

SiobhyB commented:
@dcalhoun, thank you, I really appreciate the thorough review and thoughtfulness you put into ensuring quality. 🙇‍♀️ I plan to address all of your comments, but likely won't be able to get to them all until after our upcoming meetup. I know you're going to be AFK at that point, so wanted to note here that I'll aim to find someone else to do another review after working through all the points. Thanks again for your help so far!

SiobhyB commented:
@dcalhoun, @enejb, this is a backlog item that I never managed to prioritise again, since my last activity a year ago. Weighing up [some of the hesitation](https://github.com/WordPress/gutenberg/pull/33210/files#r1001078405) around the changes and the fact that we haven't actively sought to prioritise this since June, 2021, I'm thinking of closing this PR. What are your thoughts? If we agree, I'd plan to review the accessibility improvements and spin them out into a separate PR.

dcalhoun commented:
> [...] I'm thinking of closing this PR. What are your thoughts? If we agree, I'd plan to review the accessibility improvements and spin them out into a separate PR.

My concern regarding [composition and alignment](https://github.com/WordPress/gutenberg/pull/33210#discussion_r1001066558) with the web, [unused code](https://github.com/WordPress/gutenberg/pull/33210#discussion_r1001088995), and lingering [accessibility issues](https://github.com/WordPress/gutenberg/pull/33210#discussion_r1001118330) remain for me. So, closing the PR and focusing on a subset of the changes is sensible from my perspective. 

If needed in the future, we could always reopen the PR.

SiobhyB commented:
Thank you, @dcalhoun! I'll go ahead to close this and review the subset of changes we can keep as part of my planned maintenance week next week.
